pkg/alertmanager/alertmanager.go

```diff
@@ -0,0 +1,55 @@
+package alertmanager
+
+import (
+	"context"
+	"log"
+	"time"
+
+	"github.com/prometheus/alertmanager/client"
+	"github.com/prometheus/client_golang/api"
+)
+
+type Alertmanager struct {
+	Endpoints []string
+	Expiry    time.Duration
+}
+
+func (a Alertmanager) Alert(alert Alert) error {
+	clientAlert := alert.clientAlert()
+	clientAlert.StartsAt = time.Now()
+	clientAlert.EndsAt = time.Now().Add(a.Expiry)
+	return a.push(clientAlert)
+}
+
+func (a Alertmanager) Resolve(alert Alert) error {
+	clientAlert := alert.clientAlert()
+	clientAlert.StartsAt = time.Now()
+	clientAlert.EndsAt = time.Now()
+	return a.push(clientAlert)
+}
+
+func (a Alertmanager) push(alert client.Alert) error {
+	var err error
+	var pushes int
+	for _, endpoint := range a.Endpoints {
+		apiClient, err := api.NewClient(api.Config{Address: endpoint})
+		if err != nil {
+			log.Println(err)
```

Comment:
:question: Is it okay not to exit here (e.g. `log.Fatalf`)? Will it be catched somewhere?

Reply:
Okay I've found... https://github.com/errm/alertdog/pull/1/files#diff-8e36fbeb7a03da16fee1b491a126581d93b5dec3b8bd4e4d4398aed38232bbacR52

However, in the current implementation, AFAIU, the only last `error` will be returned...? For exapmle, if we have three endpoints and two of three have errors, the first error will be logged but not catched. Is it acceptable by design, or am I missing something?

Reply:
Yeah I am trying to follow roughly the same symantics as https://github.com/prometheus/prometheus/blob/main/notifier/notifier.go#L451-L545

> It returns true if the alerts could be sent successfully to at least one Alertmanager.

I think it's fine to just return the last error, as this is mainly just a signal to trigger sending an alert to pagerduty.

If the last push suceeds then err will be nil, but then we wouldn't return an error anyway since at least 1 push was OK...

If all the pushes fail then the last one will be an error, and it's also fine to return the last error!

Perhaps returning the error from the client isn't the best design here and we should just return our own error!